tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,223 @@
+
+    def get(self, key):
+        obs = asyncio.run(self.store.get(key, prototype=default_buffer_prototype()))
+        return obs
+
+    def get_partial_values(self, key_ranges):
+        obs_partial = asyncio.run(
+            self.store.get_partial_values(
+                prototype=default_buffer_prototype(), key_ranges=key_ranges
+            )
+        )
+        return obs_partial
+
+    def delete(self, path):  # path is key
+        return asyncio.run(self.store.delete(path))
+
+    def empty(self):
+        return asyncio.run(self.store.empty())
+
+    def clear(self):
+        return asyncio.run(self.store.clear())
+
+    def exists(self, key):
+        return asyncio.run(self.store.exists(key))
+
+    def list_dir(self, prefix):
+        raise NotImplementedError
+
+    def list_prefix(self, prefix: str):
+        raise NotImplementedError
+
+    def set_partial_values(self, key_start_values):
+        raise NotImplementedError
+
+    def supports_listing(self):
+        raise NotImplementedError
+
+    def supports_partial_writes(self):
+        raise NotImplementedError
+
+    def supports_writes(self):
+        raise NotImplementedError
+
+
+class ZarrStoreStateMachine(RuleBasedStateMachine):
+    keys_bundle = Bundle("keys_bundle")
+    key_ranges_bundle = Bundle("key_ranges")
+
+    def __init__(self):
+        super().__init__()
+        self.model: dict[str, bytes] = {}
+        self.store = SyncStoreWrapper(MemoryStore(mode="w"))
+
+    # NOTE: set this and the next as initialize instead of rule to make sure they run
+    #      and create key, key_ranges before any rules that use key, key_ranges
+    #      not totally sure this is necessary
+    @initialize(key=paths, data=st.binary(min_size=0, max_size=100), target=keys_bundle)
+    def set(self, key: str, data: bytes) -> None:
+        note(f"(set) Setting {key!r} with {data}")
+        assert not self.store.mode.readonly
+        data_buf = Buffer.from_bytes(data)
+        self.store.set(key, data_buf)
+        self.model[key] = data_buf
+        note("init key")
+        return key
+
+    @initialize(key=keys_bundle, data=st.data(), target=key_ranges_bundle)
+    def init_key_ranges(self, data, key):
+        note("init key ranges")
+        key_st = st.just(key)
+        key_ranges_st = key_ranges(keys=key_st)
+        return data.draw(key_ranges_st)
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(key=keys_bundle)
+    def get(self, key) -> None:
+        note(f"(get) {key=}")
+        store_value = self.store.get(key)
+        # to bytes here necessary (on model and store) because data_buf set to model in set()
+        assert self.model[key].to_bytes() == (store_value.to_bytes())
+
+    @precondition(lambda self: len(self.model.keys()) > 0)
+    @rule(key=keys_bundle, key_range=key_ranges_bundle)
+    def get_partial_values(self, key, key_range) -> None:
+        note(f"(get partial) {key_range=}")
```

Comment:
```suggestion
    @rule()
    def get_partial_values(self) -> None:
        key_range = data.draw(key_ranges(keys=st.sampled_from(sorted(self.model.keys())))
```

let's chat about this.

Reply:
okay, that would be great. I'm currently having trouble getting it to work in conjunction with `clear()` , having delete consume the keys_bundle works fine, but i haven't figured out how to delete the entire keys bundle to match the behavior of `.clear()`